DBus Config Management (try-improvement)

slyon opened the pull request:
## Description
This implements the `io.netplan.Netplan /io/netplan/Netplan io.netplan.Netplan Config` method, which creates new `/io/netplan/Netplan/config/<ID>` objects, providing the following methods:
* `io.netplan.Netplan.Config Get`
* `io.netplan.Netplan.Config Set`
* `io.netplan.Netplan.Config Try`
* `io.netplan.Netplan.Config Apply`
* `io.netplan.Netplan.Config Cancel`
And also the `io.netplan.Netplan.Config Changed` signal.

Each new config object will copy the current global netplan state (from `/etc/netplan`) to `/tmp/netplan-config-<ID>` then Get/Set can be called on that temp directory. On Apply/Try it will backup the current global config to `/tmp/netplan-config-BACKUP` and copy over the current config state to `/etc/netplan`. On revert the data from `/tmp/netplan-config-BACKUP` will be restored.

## Checklist

- [x] Runs `make check` successfully.
- [x] Retains 100% code coverage (`make check-coverage`).
- [ ] New/changed keys in YAML format are documented.
- [ ] \(Optional\) Adds example YAML for new feature.
- [ ] \(Optional\) Closes an open bug in Launchpad.

slyon commented:
@sil2100 I'd like to ask for your review here. This got big again, sorry for that. I'm adding lots of new tests, as I was able to enable code coverage for all of the dbus code, which was previously excluded.

sil2100 reviewed (commented):
Ok, so this is not yet a full review, but I might need some clarification before proceeding. Sorry it's taking so long - it's a bit of code and it's hard to get back to after dealing with context switching.

I left two smaller inline comments so far, but I actually have one bigger thing that I wanted to make sure about here, in the main comment.

Looking at the DBus code so far, I am wondering about the DBus methods that are now available. So this PR adds the required `io.netplan.Netplan.Config` `Get()`, `Set()`, `Apply()` etc., as per the spec. And this feels fine. But then, what about the `io.netplan.Netplan` `Get()`, `Set()` methods? The ones in `netplan_vtable[]` handled by `method_{get,set}`? It feels to me right now we're offering two separate interfaces for similar things? Should we deprecate the less-specific Get/Set, or is there some use case for those that I don't remember about?

Apologies if this is a stupid question and/or if I'm mixing something up. I'll continue the review tomorrow afterwards.

slyon commented on src/dbus.c:
```diff
@@ -129,122 +232,154 @@ static int method_info(sd_bus_message *m, void *userdata, sd_bus_error *ret_erro
+        return sd_bus_error_setf(ret_error, SD_BUS_ERROR_FAILED, "cannot extract config_delta or origin_hint"); // LCOV_EXCL_LINE
 
     if (!!strcmp(origin_hint, ""))
         origin = g_strdup_printf("--origin-hint=%s", origin_hint);
```
`origin` is declared as `g_autofree gchar *origin = NULL` above, so it should be auto-free'ed after the function returns.

slyon commented on src/dbus.c:
```diff
@@ -82,14 +98,97 @@ _try_accept(bool accept, sd_bus_message *m, NetplanData *d, sd_bus_error *ret_er
+        g_object_unref(dest);
+        g_free(dest_path);
+    }
+    globfree(&gl);
```
Good point!

slyon commented:
> Ok, so this is not yet a full review, but I might need some clarification before proceeding. Sorry it's taking so long - it's a bit of code and it's hard to get back to after dealing with context switching.

Please take your time.

> I left two smaller inline comments so far, but I actually have one bigger thing that I wanted to make sure about here, in the main comment.

I fixed / commented on those.

> Looking at the DBus code so far, I am wondering about the DBus methods that are now available. So this PR adds the required `io.netplan.Netplan.Config` `Get()`, `Set()`, `Apply()` etc., as per the spec. And this feels fine. But then, what about the `io.netplan.Netplan` `Get()`, `Set()` methods? The ones in `netplan_vtable[]` handled by `method_{get,set}`? It feels to me right now we're offering two separate interfaces for similar things? Should we deprecate the less-specific Get/Set, or is there some use case for those that I don't remember about?
> 
> Apologies if this is a stupid question and/or if I'm mixing something up. I'll continue the review tomorrow afterwards.

Yes, you are right. The methods from `io.netplan.Netplan` (Get/Set/Try/Cancel) are basically the "generic" implementation, which I did as an intermediate step, wrapping `netplan get/set/try` into DBus. They are not part of the spec.

I was already thinking about deprecating them, as they were not request. But as almost all of their code is used by the methods from `io.netplan.Netplan.Config`, I thought it wouldn't hurt to keep them. But apparently they seem to confuse, so I pushed a commit to disable/comment them, for now. If you want I can also completely remove that code an refactor the other functions to execute that code directly instead via the `method_get/set/try/cancel` functions.
I do not want you to re-read and think about all that code again.. so maybe we should rather do it in a next step, after this PR?

sil2100 reviewed (approved):
Looking good from the implementation POV! There's a lot of code so maybe some things got missed, but with such code coverage I think we're really safe.

A few design-related things before we proceed with a final +1:
 * As for removing the old DBus APIs - I'm fine with leaving them commented out for now. In the future I'd prefer removing them as I think we should strive for a consistent API where there's only one way to do a get/set/apply operation not to confuse others. I mean, the implementation can even stay as is, but I would just remove the commented out sections. Of course, as you proposed, I'm fine with doing it in some follow up PRs.
 * hm, a quick question about one thing regarding the actual get/set spec compliance. The spec from the snapd team mentions something like:

> Using the object path based configuration the dbus daemon can track what configuration attempts are clobbering each other and error. If e.g. process A gets the config with id <1> and B gets it with <2> and Set() is called on <1> this automatically invalidates the other config <2> and any attempt to apply it will result in an error.

Is that actually implemented right now? I couldn't see this, as I think there is no validation during any set operations. I do see the guard for an apply when there's already another `try` happening from a different config, which might be fine, but the spec mentions a bit earlier invalidation. But it also might not be important, or might have been changed?

Anyway, +1 in overall, just give me a sign about the mentioned possible-clobbering protection.

slyon commented:
> Looking good from the implementation POV! There's a lot of code so maybe some things got missed, but with such code coverage I think we're really safe.

Thank you very much for your review of this big PR!

> * As for removing the old DBus APIs - I'm fine with leaving them commented out for now. In the future I'd prefer removing them as I think we should strive for a consistent API where there's only one way to do a get/set/apply operation not to confuse others. I mean, the implementation can even stay as is, but I would just remove the commented out sections. Of course, as you proposed, I'm fine with doing it in some follow up PRs.

ACK.
I removed the commented code and did some minimal refactoring. Some more/bigger refactoring can be done in a follow-up PR.

> * hm, a quick question about one thing regarding the actual get/set spec compliance. The spec from the snapd team mentions something like:
> 
> > Using the object path based configuration the dbus daemon can track what configuration attempts are clobbering each other and error. If e.g. process A gets the config with id <1> and B gets it with <2> and Set() is called on <1> this automatically invalidates the other config <2> and any attempt to apply it will result in an error.
> 
> Is that actually implemented right now? I couldn't see this, as I think there is no validation during any set operations. I do see the guard for an apply when there's already another `try` happening from a different config, which might be fine, but the spec mentions a bit earlier invalidation. But it also might not be important, or might have been changed?

Damn! I must have missed that part of the spec. Thanks for catching it!
The clobbering-protection is now implemented as of dca725cb452a4be29d0b0e74a42fce252aa741f2

sil2100 reviewed (approved):
Ok, I think this is good! So actually I thought I had an idea for improving the dirty-config handling, by just tracking the first `set` and resetting it whenever we apply, but then I realized that it's not that easy! And that your tracking of the invalidation makes sense. Since yes, we need to track dirtiness per `get` call as some could have been really old, from before the `set` that was first to happen, but some can be after the apply has happened. Anyway, what I wanted to say: all good! There's probably different ways we could have done this, but yours feels like a good choice.

Good work! Let me get this merged now. Thanks for working on this!

sil2100 commented:
Note for the future: we might want to sooner or later properly document the DBus API for public consumption. Let's create a card for that.
